File: frontend/src/lobby/lobby-ui.js
const h = require('snabbdom/h').default
const button = require('../common/button')
const modal = require('../common/modal')
const images = require('../common/images')
const {renderScenario} = require('../game/gameBoard/static')
const _ = require('lodash')

function render(state, actionHandler) {
  let m = null
  if (state.previewScenario) {
    m = modal(renderScenario(state.previewScenario), [actionHandler, {closeModal: true}])
  }

  return h('div',
      [h('section.hero.is-primary', [
            h('div.hero-head', h('nav.navbar', h('div.container', h('div.navbar-end', [
                  h('a.navbar-item', {props: {href: 'https://github.com/gregor-i/roboRace'}}, 'Sources @ Github')
                ]
            )))),
            h('div.hero-body',
                h('div.container', [
                  h('h1.title.is-2', 'Robo Race'),
                  h('h2.subtitle.is-3', 'Game Lobby')
                ])
            )
          ]
      ),
        renderGameList(state, state.games, actionHandler),
        renderScenarioList(state, state.scenarios, actionHandler),
        m
      ])

}

function card(content){
  return h('div.card', {style: {'marginBottom': '16px'}},
      h('div.card-content', content))
}

function mediaObject(left, content) {
  return h('article.media', [
    left ? h('figure.media-left',
        h('p.image', {style: {width: '64px', height: '64px'}},
            left
        )
    ) : null,
    h('div.media-content', content)
  ])
}

function robotImage(index, filled, you) {
  return h('img', {
    class: {
      'robot-tile': true,
      'robot-tile-you': you
    },
    props: {
      src: filled ? images.player(index) : images.playerStart(index)
    }
  })
}


function renderPlayerSlots(game) {
  return game.scenario.initialRobots.map(robot =>
      robotImage(
          robot.index,
          game.robots.find(r => r.index === robot.index),
          _.get(game.you, 'index') === robot.index
      )
  )
}

function gameState(game){
  if(game.cycle === 0)
    return 'Game waiting for players'
  else
    return 'Game running'
}

function renderGameList(state, games, actionHandler) {
  function renderGame(gameRow) {
    return card(mediaObject(null, [
      h('div', [h('strong', 'state: '), gameState(gameRow)]),
      h('div', renderPlayerSlots(gameRow)),
      button.builder.primary()(actionHandler, {openGame: gameRow}, 'Enter')
    ]))
  }

  return h('section.section',
      h('div.container', [
        h('h4.title', 'Game List: '),
        ...games.map(renderGame)
      ]))
}

function renderScenarioList(state, scenarios, actionHandler) {
  function renderScenario(scenarioRow) {
    return card(mediaObject(null,[
      h('div', [h('strong', 'Scenario description: '), scenarioRow.description]),
      h('div', scenarioRow.scenario.initialRobots.map(robot => robotImage(robot.index, false, false))),
      button.group(
            button.primary(actionHandler, {openScenario: scenarioRow}, 'Enter'),
            button.builder(actionHandler, {editScenario: scenarioRow.id}, 'Edit'),
            button.builder(actionHandler, {previewScenario: scenarioRow.scenario}, 'Preview'),
            button.builder.disabled(scenarioRow.owner !== state.player)(actionHandler, {
              deleteScenario: true,
              id: scenarioRow.id
            }, 'Delete')
        )]
    ))
  }

  return h('section.section',
      h('div.container', [
        h('h4.title', 'Scenario List: '),
        ...scenarios.map(renderScenario)
      ]))
}

module.exports = render

File: frontend/src/frontend/lobby/LobbyUi.scala
package frontend.lobby

import com.raquo.snabbdom.Modifier
import com.raquo.snabbdom.simple.{VNode, VNodeData}
import com.raquo.snabbdom.simple.events.onClick
import com.raquo.snabbdom.simple.implicits._
import com.raquo.snabbdom.simple.props.{className, disabled, href, src}
import com.raquo.snabbdom.simple.attrs.id
import com.raquo.snabbdom.simple.tags._
import frontend.Main
import frontend.common.{BulmaComponents, Images, RenderRobotImages}
import frontend.util.Ui
import gameEntities.{GameResponse, ScenarioResponse}

object LobbyUi extends Ui {
  def render(lobbyState: LobbyState): VNode =
    div(id := "robo-race",
      renderHeader(),
      seq(lobbyState.games
        .filter(_.robots.nonEmpty)
        .map(gameCard)),
      seq(lobbyState.scenarios.map(scenarioCard))
    )

  def renderHeader() =
    div(className := "hero is-primary",
      div(className := "hero-head",
        div(className := "navbar",
          div(className := "navbar-end",
            a(className := "navbar-item",
              href := "https://github.com/gregor-i/roboRace",
              "Sources @ Github"
            )
          )
        )
      ),
      div(className := "hero-body",
        div(className := "container",
          h1(className := "title is-2", "Robo Race"),
          h2(className := "subtitle is-3", "Game Lobby")
        )
      )
    )

  def gameCard(gameResponse: GameResponse): VNode = {
   val youTag: Modifier[VNode, VNodeData] = gameResponse.you match {
     case Some(you) if you.finished.exists(_.rageQuitted) => BulmaComponents.tag("Quitted", "is-danger")
     case Some(you) if you.finished.exists(!_.rageQuitted) => BulmaComponents.tag(s"Finished as ${you.finished.get.rank}", "is-primary")
     case Some(you) if you.instructionSlots.isEmpty => BulmaComponents.tag("Awaits your instructions", "is-warning")
     case _ => None
   }

    BulmaComponents.card(BulmaComponents.mediaObject(Some(RenderRobotImages(gameResponse.id.hashCode().abs % 6, filled = true)),
      div(className := "tags are-large",
        BulmaComponents.tag(s"Size: ${gameResponse.robots.size} / ${gameResponse.scenario.initialRobots.size} players", "is-info"),
        youTag,
        cond(gameResponse.robots.size < gameResponse.scenario.initialRobots.size && gameResponse.cycle == 0,
          BulmaComponents.tag("Open for new player", "is-primary"))
      )),
      Seq(
        a("Enter", onClick := (_ => Main.gotoGame(gameResponse))),
        a("Quit", onClick := (_ => println("todo")))
      )
    )
  }

  def scenarioCard(scenarioResponse: ScenarioResponse): VNode =
    BulmaComponents.card(BulmaComponents.mediaObject(Some(RenderRobotImages(scenarioResponse.id.hashCode().abs % 6, filled = true)),
      div(className := "tags are-large",
        BulmaComponents.tag(s"Description: ${scenarioResponse.description}", "is-info"),
        BulmaComponents.tag(s"Size: ${scenarioResponse.scenario.initialRobots.size} players", "is-info"),
        cond(scenarioResponse.scenario.traps.nonEmpty, BulmaComponents.tag(s"Contains traps", "is-warning")),
        cond(scenarioResponse.ownedByYou, BulmaComponents.tag(s"Created by you", "is-info"))
      )
    ), Seq(
      a("Start Game", onClick := (_ => Main.gotoPreviewScenario(scenarioResponse))),
      a("Editor", onClick := (_ => Main.gotoEditor(scenarioResponse))),
      a("Delete", onClick := (_ => println("todo"))),
    ))

}

